src/Details.js

```diff
@@ -12,55 +12,67 @@ if (typeof window !== 'undefined') {
   require('details-element-polyfill')
 }
 
-const DetailsReset = styled('details')`
+const StyledDetails = styled('details')`
   & > summary {
     list-style: none;
   }
   & > summary::-webkit-details-marker {
     display: none;
   }
+
+  ${COMMON}
 `
 function getRenderer(children) {
   return typeof children === 'function' ? children : () => children
 }
 
-function DetailsBase({children, overlay, render = getRenderer(children), defaultOpen = false, ...rest}) {
-  const [open, setOpen] = useState(defaultOpen)
+function Details({children, overlay, render = getRenderer(children), open, onToggle, defaultOpen = false, ...rest}) {
+  const [internalOpen, setInternalOpen] = useState(defaultOpen)
+  // only handle open state if user doesn't provide a value for the open prop
+  const shouldHandleOpen = typeof open === 'undefined'
   const ref = useRef(null)
 
   const closeMenu = useCallback(
     event => {
       // only close the menu if we're clicking outside
       if (event && event.target.closest('details') !== ref.current) {
-        setOpen(false)
+        setInternalOpen(false)
         document.removeEventListener('click', closeMenu)
       }
     },
     [ref]
   )
-
+  // handles the overlay behavior - closing the menu when clicking outside of it
   useEffect(() => {
-    if (overlay && open) {
+    if (shouldHandleOpen && overlay && internalOpen) {
       document.addEventListener('click', closeMenu)
       return () => {
         document.removeEventListener('click', closeMenu)
       }
     }
-  }, [open, overlay, closeMenu])
+  }, [internalOpen, shouldHandleOpen, overlay, closeMenu])
+
+  function handleToggle(e) {
+    onToggle && onToggle(e)
 
-  function toggle(event) {
-    setOpen(event.target.open)
+    if (!e.defaultPrevented) {
+      setInternalOpen(e.target.open)
+    }
   }
 
   return (
-    <DetailsReset {...rest} ref={ref} open={open} onToggle={toggle} overlay={overlay}>
+    <StyledDetails
+      {...rest}
+      ref={ref}
+      open={shouldHandleOpen ? internalOpen : open}
+      onToggle={shouldHandleOpen ? handleToggle : onToggle}
+      overlay={overlay}
+    >
       {render({open})}
-    </DetailsReset>
+    </StyledDetails>
   )
 }
```

Comment:
As this is now, if the user controls the `open` state, there's no way for them to replicate the functionality where clicking outside the component closes it unless they implement their own document click handling. As I started to dig into this, it became clear that controlling users would need to know when the "outside click" happens, just as our internal state management does. I fiddled with the component and ended up with something like this, which introduces `onClickOutside`:

```javascript
function Details({
  children,
  overlay,
  render = getRenderer(children),
  open,
  onClickOutside,
  onToggle,
  defaultOpen = false,
  ...rest
}) {
  // only handle open state if user doesn't provide a value for the open prop
  const userManagingState = typeof open !== 'undefined'
  const defaultValue = userManagingState ? open : defaultOpen
  const [internalOpen, setInternalOpen] = useState(defaultValue)
  const ref = useRef(null)
  const isOpen = open || internalOpen

  // Keep our internal state synchronized with controlled state, if any
  useEffect(() => {
    if (userManagingState && open !== internalOpen) {
      setInternalOpen(open)
    }
  }, [userManagingState, open])

  const onClickOutsideInternal = useCallback(
    event => {
      if (event.target.closest('details') !== ref.current) {
        onClickOutside && onClickOutside(event)
        if (!event.defaultPrevented) {
          setInternalOpen(false)
        }
      }
    },
    [ref, onClickOutside, setInternalOpen]
  )

  // handles the overlay behavior - closing the menu when clicking outside of it
  useEffect(() => {
    if (isOpen && overlay) {
      document.addEventListener('click', onClickOutsideInternal)
      return () => {
        document.removeEventListener('click', onClickOutsideInternal)
      }
    }
  }, [isOpen, overlay, onClickOutsideInternal])

  function handleToggle(e) {
    onToggle && onToggle(e)

    if (!e.defaultPrevented) {
      setInternalOpen(e.target.open)
    }
  }

  return (
    <StyledDetails {...rest} ref={ref} open={isOpen} onToggle={handleToggle} overlay={overlay}>
      {render({isOpen})}
    </StyledDetails>
  )
}
```

With this setup, we never toggle our internal state without also calling a user-supplied callback that might be driving a controlled state (except when we error-correct in the first effect). This allows users to fully control the component using a setup like this:

```javascript
<State default={false}>
  {([open, setOpen]) => {
    const handleToggle = (e) => setOpen(e.target.open)
    const handleClickOutside = () => setOpen(false)

    return (
      <Details open={open} onToggle={handleToggle} onClickOutside={handleClickOutside} overlay>
        <Button as="summary">Click me</Button>
        <p>This should show and hide</p>
      </Details>
    )
  }}
</State>
```

I believe the callback type would be

```typescript
onClickOutside: React.MouseEventHandler
```

Reply:
Also I'm not fully convinced the error-correcting effect is even necessary; it would only come into play if the user suddenly stopped supplying a value for `open`.

Reply:
I like this pattern! I intentionally left it up to the user to determine if an outside click happened - I figure that if someone wanted to customize the open/close behavior they will probably also want to customize how the overlay stuff behaves and figured it's not too much work to set up those click handlers themselves. But I don't think that this adds too much complexity to the API and would be nice to have so I'll add it in 🙌 

I agree - I don't think that we need to sync the user-controlled state with our state in the error correcting effect, I'll probably leave that out since it might make catching errors in state management on the user side trickier (if the open state is still working even though they aren't properly supplying an `open` prop that might be hard to troubleshoot)